Mac build using CMake fails

I want to build Aerofoil 1.1.2 from source on macOS 10.15.7 using CMake. First I create a build directory and `cd` into it:

```
mkdir build
cd build
```

Then I run `cmake`:

```
cmake ..
make
```

I get lots of warning about C++11 features being used without C++11 mode having been requested, like:

```
PortabilityLayer/CompositeRenderedFont.h:13:115: warning:
      'override' keyword is a C++11 extension [-Wc++11-extensions]
  ...int character, const GpRenderedGlyphMetrics *&outMetricsPtr, const void *&outData) const override;
                                                                                              ^
```

and then finally a failure:

```
PortabilityLayer/MenuManager.cpp:308:11: error:
      no viable conversion from returned value of type 'std::__1::nullptr_t' to function return type 'THandle<Menu>'
                        return nullptr;
                               ^~~~~~~
```

Reading CMakeLists.txt I see that C++11 mode only gets requested if `PLATFORM` is equal to `MAC`, so I try:

```
cmake -DPLATFORM=MAC ..
make
```

This fails with:

```
AerofoilX/GpMain_SDL_X.cpp:23:10: fatal error:
      'MacInit.h' file not found
#include "MacInit.h"
         ^~~~~~~~~~~
```

So I patch CMakeLists:

```diff
--- CMakeLists.txt.orig	2021-11-10 17:05:19.000000000 -0600
+++ CMakeLists.txt	2022-03-16 19:19:24.000000000 -0500
@@ -264,6 +264,7 @@
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/GpShell>
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/AerofoilSDL>
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/AerofoilPortable>
+		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/AerofoilMac/AerofoilMac>
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/PortabilityLayer>
 		${SDL2_INCLUDE_DIRS}
 		)
```

And try again:

```
cmake -DPLATFORM=MAC ..
make
```

This fails with:

```
Undefined symbols for architecture x86_64:
  "MacInit()", referenced from:
      _main in GpMain_SDL_X.cpp.o
  "PortabilityLayer::CompositeRenderedFont::CompositeRenderedFont(PortabilityLayer::RenderedFont*, PortabilityLayer::RenderedFont*)", referenced from:
      PortabilityLayer::MenuManagerImpl::GetMenuTextCompositeFont() in libPortabilityLayer.a(MenuManager.cpp.o)
  "PortabilityLayer::CompositeRenderedFont::MeasureString(unsigned char const*, unsigned long) const", referenced from:
      PortabilityLayer::MenuManagerImpl::RefreshMenuBarLayout() in libPortabilityLayer.a(MenuManager.cpp.o)
      PortabilityLayer::MenuManagerImpl::RefreshMenuLayout(Menu*) in libPortabilityLayer.a(MenuManager.cpp.o)
ld: symbol(s) not found for architecture x86_64
```

> I guess I can add a nullptr_t constructor

I see you did in d470bb5eebcf3a18ede7f3eb106563a0220deb14. Without `-DPLATFORM=MAC` the error is now:

```
PortabilityLayer/MenuManager.cpp:542:19: error: use of overloaded operator '==' is
      ambiguous (with operand types 'THandle<Menu>' and 'std::__1::nullptr_t')
                if (m_firstMenu == nullptr)
                    ~~~~~~~~~~~ ^  ~~~~~~~
PortabilityLayer/PLHandle.h:44:7: note: candidate function
        bool operator==(const THandle<T> &other) const;
             ^
PortabilityLayer/PLHandle.h:47:7: note: candidate function
        bool operator==(T** other) const;
             ^
PortabilityLayer/MenuManager.cpp:559:19: error: use of overloaded operator '==' is
      ambiguous (with operand types 'THandle<Menu>' and 'std::__1::nullptr_t')
                if (m_firstMenu == nullptr)
                    ~~~~~~~~~~~ ^  ~~~~~~~
PortabilityLayer/PLHandle.h:44:7: note: candidate function
        bool operator==(const THandle<T> &other) const;
             ^
PortabilityLayer/PLHandle.h:47:7: note: candidate function
        bool operator==(T** other) const;
             ^
PortabilityLayer/MenuManager.cpp:750:46: error: use of overloaded operator '==' is
      ambiguous (with operand types 'THandle<Menu>' and 'std::__1::nullptr_t')
                if (m_menuSelectionState.GetSelectedMenu() == nullptr)
                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~~~~~
PortabilityLayer/PLHandle.h:44:7: note: candidate function
        bool operator==(const THandle<T> &other) const;
             ^
PortabilityLayer/PLHandle.h:47:7: note: candidate function
        bool operator==(T** other) const;
             ^
```

And with `-DPLATFORM=MAC` the error is:

```
In file included from PortabilityLayer/PLHandle.cpp:1:
PortabilityLayer/PLHandle.h:27:10: error: use of undeclared identifier 'std'
        THandle(std::nullptr_t);
                ^
PortabilityLayer/PLHandle.h:79:20: error: out-of-line definition of 'THandle<T>' does not
      match any declaration in 'THandle<T>'
inline THandle<T>::THandle(std::nullptr_t)
                   ^~~~~~~
2 errors generated.
```

So I think that change does not work and needs to be reverted or needs further changes.

The `warning: 'override' keyword is a C++11 extension [-Wc++11-extensions]` warnings remain (when not `-DPLATFORM=MAC`). If you're using C++11 features like `override` why not unconditionally request C++11 support? Why do it conditionally only on macOS?

I'll submit a PR to address the rest of the issues.

--- a/AerofoilX/GpMain_SDL_X.cpp
+++ b/AerofoilX/GpMain_SDL_X.cpp
@@ -20,7 +20,9 @@
 #include "IGpVOSEventQueue.h"
 
 #include <string>
+#ifdef __MACOS__
 #include "MacInit.h"
+#endif
 
 GpXGlobals g_gpXGlobals;
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,6 +45,7 @@ add_library(PortabilityLayer STATIC
 	PortabilityLayer/BitmapImage.cpp
 	PortabilityLayer/ByteSwap.cpp
 	PortabilityLayer/CFileStream.cpp
+	PortabilityLayer/CompositeRenderedFont.cpp
 	PortabilityLayer/DeflateCodec.cpp
 	PortabilityLayer/DialogManager.cpp
 	PortabilityLayer/DisplayDeviceManager.cpp
@@ -240,7 +241,8 @@ target_include_directories(GpApp PRIVATE
 target_link_libraries(GpApp PortabilityLayer)
 
 if(CMAKE_HOST_UNIX)
-	add_executable(${EXECNAME}
+	set(EXEC_SOURCES )
+	list(APPEND EXEC_SOURCES
 		AerofoilPortable/GpSystemServices_POSIX.cpp
 		AerofoilPortable/GpThreadEvent_Cpp11.cpp
 		AerofoilPortable/GpAllocator_C.cpp
@@ -257,8 +259,16 @@ if(CMAKE_HOST_UNIX)
 		AerofoilX/GpSystemServices_X.cpp
 		AerofoilX/GpFileSystem_X.cpp
 		)
-		
-	target_include_directories(${EXECNAME} PRIVATE
+
+	set(EXEC_LIBS )
+	list(APPEND EXEC_LIBS
+		${SDL2_LIBRARIES}
+		GpApp
+		GpShell
+		)
+
+	set(EXEC_INC_DIRS )
+	list(APPEND EXEC_INC_DIRS
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/Common>
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/GpCommon>
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/GpShell>
@@ -267,8 +277,22 @@ if(CMAKE_HOST_UNIX)
 		$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/PortabilityLayer>
 		${SDL2_INCLUDE_DIRS}
 		)
+	if(PLATFORM STREQUAL "MAC")
+		list(APPEND EXEC_SOURCES
+			AerofoilMac/AerofoilMac/AerofoilApplication.mm
+			AerofoilMac/AerofoilMac/MacInit.mm
+			)
+		list(APPEND EXEC_INC_DIRS
+			$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/AerofoilMac/AerofoilMac>
+			)
+		list(APPEND EXEC_LIBS
+			"-framework Cocoa"
+			)
+	endif(PLATFORM STREQUAL "MAC")
 
-	target_link_libraries(${EXECNAME} ${SDL2_LIBRARIES} GpApp GpShell)
+	add_executable(${EXECNAME} ${EXEC_SOURCES})
+	target_include_directories(${EXECNAME} PRIVATE ${EXEC_INC_DIRS})
+	target_link_libraries(${EXECNAME} ${EXEC_LIBS})
 endif()
 
 

